src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use borrow_check::{Context, MirBorrowckCtxt};
+use borrow_check::nll::region_infer::{Cause, RegionInferenceContext};
+use dataflow::BorrowData;
+use rustc::mir::{Local, Location, Mir};
+use rustc::mir::visit::{MirVisitable, PlaceContext, Visitor};
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::DiagnosticBuilder;
+use util::liveness::{self, DefUse, LivenessMode};
+
+impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
+    pub fn explain_why_borrow_contains_point(
+        &self,
+        context: Context,
+        borrow: &BorrowData<'_>,
+        err: &mut DiagnosticBuilder<'_>,
+    ) {
+        if let Some(regioncx) = &self.nonlexical_regioncx {
+            if let Some(cause) = regioncx.why_region_contains_point(borrow.region, context.loc) {
+                let mir = self.mir;
+
+                cause.label_diagnostic(mir, err);
```

Comment:
maybe we should move this call to only execute if `cause.root_cause()` returns something other than `LiveVar` or `DropVar`... that would avoid dump extra causal information when we can give a more user-friendly version. But really I guess we just want to make the "cause-gathering" stuff enabled or lazy, so never mind.

Reply:
:+1:, so should just leave as is ?